osmosis-labs/osmosis
# XCS + Registries integration: IBC forward generation using registries

## What is the purpose of the change

This PR is contains #4683. While that one only replaced the prefix registry, this one replaces the whole IBC message creation to support forwarding. 

This is still a draft at the moment. Some things missing:

 * Pre-swap unwrapping
 * How to deal with validation and user specified memos
 * Tests specific to multiple hops
 * More/better tests
    * All chains have the osmo prefix, which makes it hard to test all potential validation paths. Changing this has proven harder than expected, so how we test may depend on our ability to change those prefixes in the testing app.
 * Add a registry to check if a chain allows forwarding. 
 * Cleanup

## Brief Changelog

 * Contracts now use the registries to build the forward package
 * Lots of refactoring

## Testing and Verifying

Tests have been adapted to the new changes.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)
